### PR TITLE
Sync parents

### PR DESCRIPTION
**Pull request by shayzluf**
[Resolves] #3665

---

# Description

**Write why you are making the changes in this pull request**
If a parent block is missing from the current block that is going to be processed. we first request it from p2p and process it before processing the orphaned block.

### CODE REVIEW

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 1]
```diff
@@ -125,19 +125,54 @@ func (s *InitialSync) roundRobinSync(genesis time.Time) error {
 		sort.Slice(blocks, func(i, j int) bool {
 			return blocks[i].Slot < blocks[j].Slot
 		})
+		var checkParentExists func(blk *eth.BeaconBlock, parentBlocks []*eth.BeaconBlock) ([]*eth.BeaconBlock, error)
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/round_robin.go]: why is this like this ? make this another pointer receiver method

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 2]
```diff
@@ -125,19 +125,54 @@ func (s *InitialSync) roundRobinSync(genesis time.Time) error {
...
 
-		for _, blk := range blocks {
-			logSyncStatus(genesis, blk, peers, counter)
-			if featureconfig.FeatureConfig().InitSyncNoVerify {
-				if err := s.chain.ReceiveBlockNoVerify(ctx, blk); err != nil {
+		var receiveBlocks func(blocks []*eth.BeaconBlock) error
+		receiveBlocks = func(blocks []*eth.BeaconBlock) error {
+			for _, blk := range blocks {
+				logSyncStatus(genesis, blk, peers, counter)
+				emptyBlk := make([]*eth.BeaconBlock, 0)
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/round_robin.go]: ? dont understand this part. whats the purpose of this

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 3]
```diff
@@ -125,19 +125,54 @@ func (s *InitialSync) roundRobinSync(genesis time.Time) error {
...
+			ok, err := s.chain.ParentExists(ctx, blk)
+			if err != nil {
+				return nil, err
+			}
+			if !ok {
+				bl, err := s.requestBlocksByRoot(ctx, [][32]byte{bytesutil.ToBytes32(blk.ParentRoot)}, peers[0])
+				if err != nil {
+					return nil, err
+				}
+				parentBlocks = append(bl, parentBlocks...)
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/round_robin.go]: you are only ever looking for 1 parent block, a slice isnt needed here

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: changed

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: changed

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: to recursively look and process missing parents

Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: i stopped using it . removing

**Comment by shayzluf** [comment]
@nisdas i took out the method into a pointer receiver. tested ParentExists
Thanks

File: beacon-chain/blockchain/receive_block.go [hunk 4]
```diff
@@ -209,3 +210,17 @@ func isCompetingBlock(root []byte, slot uint64, headRoot []byte, headSlot uint64
 		competingBlks.Inc()
 	}
 }
+
+//ParentExists checks block parent is in store.
+func (s *Service) ParentExists(ctx context.Context, block *ethpb.BeaconBlock) (bool, error) {
+	ctx, span := trace.StartSpan(ctx, "beacon-chain.blockchain.ParentExists")
+	defer span.End()
+	parentBlk, err := s.beaconDB.Block(ctx, bytesutil.ToBytes32(block.ParentRoot))
```
Comment (nisdas) [inline beacon-chain/blockchain/receive_block.go]: make this a HasBlock

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 5]
```diff
@@ -126,18 +126,35 @@ func (s *InitialSync) roundRobinSync(genesis time.Time) error {
 			return blocks[i].Slot < blocks[j].Slot
 		})
 
-		for _, blk := range blocks {
-			logSyncStatus(genesis, blk, peers, counter)
-			if featureconfig.FeatureConfig().InitSyncNoVerify {
-				if err := s.chain.ReceiveBlockNoVerify(ctx, blk); err != nil {
+		var receiveBlocks func(blocks []*eth.BeaconBlock) error
+		receiveBlocks = func(blocks []*eth.BeaconBlock) error {
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/round_robin.go]: this is dangerous, what if we never find a parent that exists in our db ?

Comment (nisdas) [reply beacon-chain/sync/initial-sync/round_robin.go]: also this method should also be a pointer receiver

Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: i can limit the depth of it. but it should satisfy both successfully retrieving a parent forever

Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: changed to pointer receiver

Comment (shayzluf) [reply beacon-chain/sync/initial-sync/round_robin.go]: limited the depth of recursive calls to 10

File: beacon-chain/blockchain/receive_block.go [hunk 6]
```diff
@@ -209,3 +210,11 @@ func isCompetingBlock(root []byte, slot uint64, headRoot []byte, headSlot uint64
 		competingBlks.Inc()
 	}
 }
+
+//ParentExists checks block parent is in store.
+func (s *Service) ParentExists(ctx context.Context, block *ethpb.BeaconBlock) bool {
```
Comment (terencechain) [inline beacon-chain/blockchain/receive_block.go]: You dont need this, InitialSync should have access to the DB, they can just call it there. I don't think BlockReceive should provide this just to call DB, It's not chain service dependent

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 7]
```diff
@@ -187,6 +179,45 @@ func (s *InitialSync) roundRobinSync(genesis time.Time) error {
 	return nil
 }
 
+func (s *InitialSync) receiveBlocks(ctx context.Context, blocks []*eth.BeaconBlock, counter *ratecounter.RateCounter, peers []peer.ID, genesis time.Time, depth uint8) error {
```
Comment (terencechain) [inline beacon-chain/sync/initial-sync/round_robin.go]: test?

File: beacon-chain/blockchain/receive_block.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/blockchain/receive_block.go]: you are right. moving it

**Comment by shayzluf** [comment]
Ok, will revert and fix the test

**Pull request closed without merging**